Storage/tests/Unit/SigningHelperTest.php

```diff
@@ -789,6 +791,68 @@ private function mockConnection($credentials = null, $rw = null)
 
         return $conn->reveal();
     }
+
+    public function testRetrySignBlobSuccessFirstAttempt()
+    {
+        $signBlobFn = function () {
+            return 'signature';
+        };
+
+        $res = $this->helper->proxyPrivateMethodCall('retrySignBlob', [
+            $signBlobFn
+        ]);
+
+        $this->assertEquals('signature', $res);
+    }
+
+    public function testRetrySignBlobSuccessAfterRetries()
+    {
+        $attempts = 0;
+        $signBlobFn = function () use (&$attempts) {
+            $attempts++;
+            if ($attempts < 5) {
+                throw new ServiceException('Transient error', 503);
+            }
+            return 'signature';
+        };
+
+        $res = $this->helper->proxyPrivateMethodCall('retrySignBlob', [
+            $signBlobFn
+        ]);
+
+        $this->assertEquals('signature', $res);
+        $this->assertEquals(5, $attempts);
+    }
+
+    public function testRetrySignBlobNonRetryableError()
+    {
+        $this->expectException(ServiceException::class);
+        $this->getExpectedExceptionMessage('Non-retryable error');
+
+        $signBlobFn = function () {
+            throw new ServiceException('Non-retryable error', 400);
+        };
+
+        $res = $this->helper->proxyPrivateMethodCall('retrySignBlob', [
+            $signBlobFn
+        ]);
+
+        $this->assertEquals('Non-retryable error', $res);
+    }
+
+    public function testRetrySignBlobRetriesExhausted()
+    {
+        $this->expectException(\RuntimeException::class);
+        $this->getExpectedExceptionMessage('Failed to sign message after maximum attempts.');
+
+        $signBlobFn = function () {
+            throw new ServiceException('Transient error', 503);
+        };
+
+        $res = $this->helper->proxyPrivateMethodCall('retrySignBlob', [
```

Comment:
generally speaking, calling private methods is a bad way to test. It would be better to mock the `Connection` class being passed to `v2Sign`, in a way that can trigger the retry.

I can work with you on how to write tests like that, if you'd like.

Reply:
That's a great point. Mocking the Connection class to trigger retries is a much more robust and testable approach. I'd be happy to work with you on implementing this.